File: triangle/triangle.py
class Solution:
    # [0] finish
    # [0, 1] ^
    # [0, 1, 2] ^
    # [0, 1, 2, 3] ^ start
    # [0, 1, 2, 3, 4]
    
    def minimumTotal(self, triangle: list[list[int]]) -> int:
        for row in range(len(triangle)-2, -1, -1): # Start from bottom
            for col in range(len(triangle[row])-1, -1, -1): # Work backwards thru rows
                triangle[row][col] += min(triangle[row+1][col], triangle[row+1][col+1]) # Dynamically allocate data
        return triangle[0][0]

File: triangle/triangle.java
class Solution 
{
    public int minimumTotal(List<List<Integer>> triangle) 
    {
        // Setup a list to store minimum paths taken
        List<Integer> minPaths = triangle.get(triangle.size()-1);
        
        // Starting from the bottom of triangle, work up finding the minimum path
        for (int level = triangle.size()-2; level > -1; level--)
            for (int i = 0; i <= level; i++)
                minPaths.set(i, Math.min(minPaths.get(i), minPaths.get(i+1)) + triangle.get(level).get(i));

        return minPaths.get(0);
    }
}